File: react-app/src/store/aircraft.js
const SET_AIRCRAFT = 'aircraft/SET_AIRCRAFT'
const SET_AIRCRAFTS = 'aircraft/SET_AIRCRAFTS'
const REMOVE_AIRCRAFT = 'aircraft/REMOVE_AIRCRAFT'

const setAircraft = (aircraft) => ({
  type: SET_AIRCRAFT,
  aircraft
})

const setAircrafts = (aircraft) => ({
  type: SET_AIRCRAFTS,
  aircraft
})

const removeAircraft = (id) => ({
  type: REMOVE_AIRCRAFT,
  id
})

export const getAircraft = () => async dispatch => {
  const res = await fetch('/api/aircraft/');

  if (res.ok) {
    const Aircraft = await res.json();
    dispatch(setAircraft(Aircraft))
  }
}


export const getAircrafts = (id) => async dispatch => {
  const res = await fetch(`/api/aircraft/${id}`)

  if (res.ok) {
    const Aircraft = await res.json();
    dispatch(setAircrafts(Aircraft))
  }
}

export const createAircraft = (aircraft) => async dispatch => {
    console.log('in route')
  const { user_id, price, manufacturer,
    name, description, cover_img, avionics,
    ifr_cert, need_IR, need_CSEL, need_CMEL, need_ATP, need_CFI,
    need_CFII, need_MEI, need_complex, need_performance,
    airport, type, gph, fuel_capacity, cruise_speed, usable_load,
    seats, poh } = aircraft;

  const res = await fetch('/api/aircraft/', {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
        user_id, price, manufacturer,
        name, description, cover_img, avionics,
        ifr_cert, need_IR, need_CSEL, need_CMEL, need_ATP, need_CFI,
        need_CFII, need_MEI, need_complex, need_performance,
        airport, type, gph, fuel_capacity, cruise_speed, usable_load,
        seats, poh
    })
  });

  if (res.ok) {
    const data = await res.json();
    dispatch(setAircrafts(data))
    return data
  } else if (res.status < 500) {
    const data = await res.json();
    if (data.errors) {
      return data.errors;
    }
  } else {
    return ['An error occurred. Please try again.']
  }

}

// export const updateEvent = (event) => async dispatch => {
//   const { name, user_id, category,
//     description, address, city, state,
//     image, start, end } = event;

//   const res = await fetch(`/api/events/${event.id}`, {
//     method: "PUT",
//     headers: {
//       "Content-Type": "application/json"
//     },
//     body: JSON.stringify({
//       name, user_id, category,
//       description, address, city, state,
//       image, start, end
//     })
//   });

//   if (res.ok) {
//     const data = await res.json();
//     console.log('monkeydata', data)
//     dispatch(setEvent(data))
//     return data
//   } else if (res.status < 500) {
//     const data = await res.json();
//     if (data.errors) {
//       return data.errors;
//     }
//   } else {
//     return ['An error occurred. Please try again.']
//   }
// }

// export const deleteEvent = (id) => async dispatch => {
//   console.log('inside thunk start', id)
//   const res = await fetch(`/api/events/${id}`, {
//     method: "DELETE",
//   });

//   if (res.ok) {
//     console.log('res.ok passes')
//     await res.json();
//     console.log('after res ok', res.json)
//     dispatch(removeEvent(id))
//   }
// }

const initialState = {};

const aircraftReducer = (state = initialState, action) => {
  let newState = { ...state };
  switch (action.type) {
    case SET_AIRCRAFT:
      action.aircraft.aircraft.forEach(air => {
        newState[air.id] = air
      })
      return newState
    case SET_AIRCRAFTS:
      action.aircraft.aircraft.forEach(air => {
        newState[air.id] = air
      })
      return newState
    case REMOVE_AIRCRAFT:
      delete newState[action.id]
      return newState
    default:
      return state;
  }
}

export default aircraftReducer

File: app/models/aircraft.py
from .db import db
import simplejson as json

class Aircraft(db.Model):
    __tablename__ = 'aircraft'

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer,  db.ForeignKey('users.id'), nullable=False)
    price = db.Column(db.Integer)
    manufacturer = db.Column(db.String(100), nullable=False)
    name = db.Column(db.String(100), nullable=False)
    year = db.Column(db.Integer, nullable=False)
    tail_number = db.Column(db.String(15), nullable=False)
    description = db.Column(db.String(100), nullable=False)
    cover_img = db.Column(db.String(), nullable=False)
    avionics = db.Column(db.String, nullable=False)
    ifr_cert = db.Column(db.Boolean)
    need_IR = db.Column(db.Boolean)
    need_CSEL = db.Column(db.Boolean)
    need_CMEL = db.Column(db.Boolean)
    need_ATP = db.Column(db.Boolean)
    need_CFI = db.Column(db.Boolean)
    need_CFII = db.Column(db.Boolean)
    need_MEI = db.Column(db.Boolean)
    need_complex = db.Column(db.Boolean)
    need_performance = db.Column(db.Boolean)
    airport = db.Column(db.String, nullable=False)
    type = db.Column(db.String)
    gph = db.Column(db.Integer)
    fuel_capacity = db.Column(db.Integer)
    cruise_speed = db.Column(db.Integer)
    usable_load = db.Column(db.Integer)
    seats = db.Column(db.Integer)
    poh = db.Column(db.String)

    comments = db.relationship("Comment", back_populates='aircraft', cascade='delete')
    user = db.relationship("User", back_populates="aircraft")
    flights = db.relationship('Flight', back_populates='aircraft', cascade='delete')
    images = db.relationship('Image', back_populates='aircraft', cascade='delete')

    def to_dict(self):
        return {
            'id': self.id,
            'user_id': self.user_id,
            'price': self.price,
            'manufacturer': self.manufacturer,
            'name': self.name,
            'year': self.year,
            'tail_number': self.tail_number,
            'description': self.description,
            'cover_img': self.cover_img,
            'avionics': self.avionics,
            'ifr_cert': self.ifr_cert,
            'need_IR': self.need_IR,
            'need_CSEL': self.need_CSEL,
            'need_CMEL': self.need_CMEL,
            'need_ATP': self.need_ATP,
            'need_CFI': self.need_CFI,
            'need_CFII': self.need_CFII,
            'need_MEI': self.need_MEI,
            'need_complex': self.need_complex,
            'need_performance': self.need_performance,
            'airport': self.airport,
            'type': self.type,
            'gph': self.gph,
            'fuel_capacity': self.fuel_capacity,
            'cruise_speed': self.cruise_speed,
            'usable_load': self.usable_load,
            'seats': self.seats,
            'poh': self.poh,
            'images': [image.to_dict() for image in self.images]
        }
